src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs

```diff
@@ -529,6 +533,86 @@ protected override void DecodeWellKnownAttributeImpl(ref DecodeWellKnownAttribut
             }
         }
 
+#nullable enable
+
+        internal bool RequiresRefSafetyRulesAttribute()
+        {
+            if (_lazyRequiresRefSafetyRulesAttribute == ThreeState.Unknown)
+            {
+                bool value = UseUpdatedEscapeRules && namespaceIncludesRefs(GlobalNamespace);
+                _lazyRequiresRefSafetyRulesAttribute = value.ToThreeState();
+            }
+            return _lazyRequiresRefSafetyRulesAttribute.Value();
+
+            static bool namespaceIncludesRefs(NamespaceSymbol ns)
+            {
+                foreach (var member in ns.GetMembersUnordered())
+                {
+                    switch (member.Kind)
+                    {
+                        case SymbolKind.Namespace:
+                            if (namespaceIncludesRefs((NamespaceSymbol)member))
+                            {
+                                return true;
+                            }
+                            break;
+                        case SymbolKind.NamedType:
+                            if (typeIncludesRefs((NamedTypeSymbol)member))
+                            {
+                                return true;
+                            }
+                            break;
+                        default:
+                            throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                    }
+                }
+                return false;
+            }
+
+            static bool typeIncludesRefs(NamedTypeSymbol type)
```

Comment:
Consider using `TypeSymbolExtensions.VisitType`

Reply:
It's not clear to me where we'd use `VisitType()`.